File: 단계별로 풀어보기/정렬/2108/통계학.cpp
#include <iostream>
#include <cmath>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int N, num, median;
    int m = 0;
    int nums[8001] = {0, };
    cin >> N;
    
    cin >> num;
    int mean = num;
    int mode;
    int mode1 = 0;
    bool second;
    nums[num+4000]++;
    
    int maximum = num;
    int minimum = num;
    
    for(int i = 1;  i < N; i++){
        cin >> num;
        mean += num;
        
        nums[num+4000]++;
        
        if (maximum < num)
            maximum = num;
        
        if (minimum > num)
            minimum = num;
    }
    
    for(int i = 0; i <= 8000; i++){
        if(nums[i] > mode1){
            mode = i;
            mode1 = nums[mode];
            second = false;
        } else if(nums[i] == mode1 && !second){
            second = true;
            mode = i;
        }
        
        while (nums[i] != 0){
            nums[i]--;
            m++;
            if(m == N/2+1){
                median = i-4000;
            }
        }
    }
    
    cout << (int)(round((double)mean/(double)N)) << '\n' << median << '\n'<< mode-4000 << '\n' << maximum - minimum;
    
    return 0;
}

File: 단계별로 풀어보기/정렬/2108/통계학.py
import sys
    
def init():
    N = int(sys.stdin.readline())
    nums = []
    for i in range(8001):
        nums.append(0)

    num = int(sys.stdin.readline())
    nums[num+4000] += 1
    total = num
    maximum = num
    minimum = num
    mode1 = 0
    cnt = 0
    second = False
    
    for i in range(N-1):
        num = int(sys.stdin.readline())
        total += num
        nums[num+4000] += 1
        
        if num > maximum:
            maximum = num
            
        if num < minimum:
            minimum = num

    for i in range(8001):
        if nums[i] > mode1:
            mode1 = nums[i]
            mode = i-4000
            second = False
        elif (nums[i] == mode1) and (not second):
            second = True
            mode = i-4000

        while nums[i] != 0:
            nums[i] -= 1
            cnt += 1
            if cnt == N//2+1:
                median = i-4000

    
    sys.stdout.write('{}\n{}\n{}\n{}' .format(round(total/N), median, mode, maximum-minimum))
        
init()
